source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRViewController.mm

```diff
@@ -149,4 +170,212 @@ - (void)render
+             // generate key for imageMap from image element's id
+             NSString *key = [NSString stringWithCString:imgElem->GetId().c_str() encoding:[NSString defaultCStringEncoding]];
+             NSURL *url = [NSURL URLWithString:urlStr];
+             // download image
+             UIImage *img = [UIImage imageWithData:[NSData dataWithContentsOfURL:url]];
+             CGSize cgsize = [ACRImageRenderer getImageSize:imgElem->GetImageSize() withHostConfig:_hostConfig];
+             // scale image
+             UIGraphicsBeginImageContext(cgsize);
+             [img drawInRect:(CGRectMake(0, 0, cgsize.width, cgsize.height))];
+             img = UIGraphicsGetImageFromCurrentImageContext();
+             UIGraphicsEndImageContext();
+             // UITask can't be run on global queue, add task to main queue
+             dispatch_async(dispatch_get_main_queue(),
+                 ^{
+                      __block UIImageView *view = nil;
+                      // syncronize access to image map
+                      dispatch_sync(_serial_queue,
+                          ^{
+                               if(!_imageViewMap[key]) {// UIImageView is not ready, cashe UIImage
+                                   _imageViewMap[key] = img;
+                               } else {// UIImageView ready, get view
+                                   view = _imageViewMap[key];
+                               }
+                          });
+                      // if view is available, set image to it, and continue image processing
+                      if(view) {
+                          view.image = img;
+                          view.contentMode = UIViewContentModeScaleAspectFit;
+                          view.clipsToBounds = NO;
+                          if(imgElem->GetImageStyle() == ImageStyle::Person) {
+                              CALayer *imgLayer = view.layer;
+                              [imgLayer setCornerRadius:cgsize.width/2];
+                              [imgLayer setMasksToBounds:YES];
+                          }
+                          // remove tag
+                          std::string id = imgElem->GetId();
+                          std::size_t idx = id.find_last_of('_');
+                          imgElem->SetId(id.substr(0, idx));
+                      }
+                  });
+             }
+    );
+}
+// add postfix to existing BaseCardElement ID to be used as key
+-(void) tagBaseCardElement:(std::shared_ptr<BaseCardElement> const &) elem
+{
+    std::string serial_number_as_string = std::to_string(_serialNumber);
+    // ID field is optional, if empty, add new one
+    if("" == elem->GetId()) {
+        elem->SetId("_" + serial_number_as_string);
+    } else { // concat a newly generated key to a existing id, the key will be removed after use
+        elem->SetId(elem->GetId() + "_" + serial_number_as_string);
```

Comment:
You don't need the if statement. Just use this line all the time.

Reply:
thanks

---
In reply to: [168014387](https://github.com/Microsoft/AdaptiveCards/pull/1124#discussion_r168014387) [](ancestors = 168014387)